scylla-cdc-printer/src/printer.rs

```diff
@@ -0,0 +1,246 @@
+use std::sync::Arc;
+use std::time::SystemTime;
+
+use anyhow;
+use chrono::Duration;
+use core::time;
+use scylla::Session;
+use scylla::frame::response::result::Row;
+use tokio::time::sleep;
+
+use scylla_cdc::cdc_types::GenerationTimestamp;
+use scylla_cdc::reader::StreamReader;
+use scylla_cdc::stream_generations::GenerationFetcher;
+
+struct CDCLogPrinter {
+    session: Arc<Session>,
+    start_timestamp: Duration,
+    end_timestamp: Duration,
+    sleep_interval: time::Duration,
+    window_size: Duration,
+    safety_interval: Duration,
+    readers: Vec<Arc<StreamReader>>,
+}
+
+impl CDCLogPrinter {
+    fn new(
+        session: &Arc<Session>,
+        start_timestamp: Duration,
+        end_timestamp: Duration,
+        window_size: Duration,
+        safety_interval: Duration,
+        sleep_interval: time::Duration,
+    ) -> CDCLogPrinter {
+        CDCLogPrinter {
+            session: Arc::clone(session),
+            start_timestamp,
+            end_timestamp,
+            window_size,
+            safety_interval,
+            sleep_interval,
+            readers: vec![],
+        }
+    }
+
+    pub async fn run(
+        &mut self,
+        keyspace: String,
+        table_name: String,
+        mut after_fetch_callback: impl FnMut(Row),
```

Comment:
Working with a single `after_fetch_callback` is not ideal because it needs exclusive access in order to be called - you would have to put it into an `Arc<Mutex<CallbackType>>` and acquire a mutex before passing each row to it which is inefficient.

I'd recommend using the `ConsumerFactory` and `Consumer` framework we already have - the printer would take a factory and create separate consumers for each of the tokio tasks.

(Besides, a printer taking an arbitrary callback does not look like a printer to me - more like a minimal version of the library)

Reply:
You can write a test factory which takes a callback and returns consumers that use the callback internally. The consumers would share the callback via `Arc<Mutex<Callback>>`.

Here is a snippet of code to make it more clear what I mean:

```rust
struct TestConsumerFactory {
    cb: Arc<Mutex<dyn FnMut(CDCRow)>>,
}

impl ConsumerFactory for TestConsumerFactory {
    fn new_consumer(&self) -> Box<dyn Consumer> {
        Box::new(TestConsumer { cb: Arc::clone(&self.cb) }
    }
}

struct TestConsumer {
    cb: Arc<Mutex<dyn FnMut(CDCRow)>>,
}

#[async_trait]
impl Consumer for TestConsumer {
    async fn consume_cdc(&mut self, data: CDCRow) {
        (self.cb.lock().unwrap())(data);
    }
}
```

Although perhaps using the factory + consumers now may be an overkill for the printer and its tests, they will be required for the minimal version of the library, so if you do it now you will have less work later.